File: examplesForThesis-R/ch4-example-gp-regression.R
########################################################################
#
# Code to reproduce examples from the Licentiate's thesis:
# Sequential Monte Carlo for inference in nonlinear state space models
#
# written by
# Johan Dahlin ( johan.dahlin (at) liu.se
#
# Available from:
# http://users.isy.liu.se/en/rt/johda87/
#
# Copyright (c) 2014 Johan Dahlin
# 
# Run this code together with the corresponding R-file to reproduce:
#
# GP regression
# Example 4.6 in Section 4.4.1
#
########################################################################

dat  = read.table("ch4-example-gp-regression-data.csv",sep=",",header=T)[,-1]
datr = read.table("ch4-example-gp-regression-datared.csv",sep=",",header=T)[,-1]
rbfa  = read.table("ch4-example-gp-regression-rbf-10d.csv",sep=",",header=T)[,-1]
m32a  = read.table("ch4-example-gp-regression-m32-10d.csv",sep=",",header=T)[,-1]
m52a  = read.table("ch4-example-gp-regression-m52-10d.csv",sep=",",header=T)[,-1]
rbfb  = read.table("ch4-example-gp-regression-rbf-20d.csv",sep=",",header=T)[,-1]
m32b  = read.table("ch4-example-gp-regression-m32-20d.csv",sep=",",header=T)[,-1]
m52b  = read.table("ch4-example-gp-regression-m52-20d.csv",sep=",",header=T)[,-1]

layout(matrix(1:6, 3, 2, byrow = FALSE)); par(mar=c(4,5,1,1)) 

plot(dat[,1],rbfa[,1],type="l",lwd=3,ylim=c(-8,30),xlim=c(0.30,9.65),xlab="x",ylab="f(x)",col="darkgreen"); 
polygon( c(dat[,1],rev(dat[,1])), c(rbfa[,3], rev(rbfa[,2])), col="lightgrey", border=NA)
lines(dat[,1],rbfa[,1],lwd=3,col="darkgreen"); lines(dat[,1],dat[,2],lwd=3,lty="dashed"); points(datr[1:5,1],datr[1:5,2],pch=19,cex=1.5)
text(10,30,"SE, N=5",pos=2);

plot(dat[,1],m52a[,1],type="l",lwd=3,ylim=c(-8,30),xlim=c(0.30,9.65),col="darkred",xlab="x",ylab="f(x)"); 
polygon( c(dat[,1],rev(dat[,1])), c(m52a[,3], rev(m52a[,2])), col="lightgrey", border=NA)
lines(dat[,1],m52a[,1],lwd=3,col="darkred"); lines(dat[,1],dat[,2],lwd=3,lty="dashed"); points(datr[1:5,1],datr[1:5,2],pch=19,cex=1.5)
text(10,30,"Matérn 5/2, N=5",pos=2);

plot(dat[,1],m32a[,1],type="l",lwd=3,ylim=c(-8,30),xlim=c(0.30,9.65),col="steelblue",xlab="x",ylab="f(x)"); 
polygon( c(dat[,1],rev(dat[,1])), c(m32a[,3], rev(m32a[,2])), col="lightgrey", border=NA)
lines(dat[,1],m32a[,1],lwd=3,col="steelblue"); lines(dat[,1],dat[,2],lwd=3,lty="dashed"); points(datr[1:5,1],datr[1:5,2],pch=19,cex=1.5)
text(10,30,"Matérn 3/2, N=5",pos=2);

plot(dat[,1],rbfb[,1],type="l",lwd=3,ylim=c(-8,30),xlim=c(0.30,9.65),xlab="x",ylab="f(x)",col="darkgreen"); 
polygon( c(dat[,1],rev(dat[,1])), c(rbfb[,3], rev(rbfb[,2])), col="lightgrey", border=NA)
lines(dat[,1],rbfb[,1],lwd=3,col="darkgreen"); lines(dat[,1],dat[,2],lwd=3,lty="dashed"); points(datr[1:15,1],datr[1:15,2],pch=19,cex=1.5)
text(10,30,"SE, N=15",pos=2);

plot(dat[,1],m52b[,1],type="l",lwd=3,ylim=c(-8,30),xlim=c(0.30,9.65),col="darkred",xlab="x",ylab="f(x)"); 
polygon( c(dat[,1],rev(dat[,1])), c(m52b[,3], rev(m52b[,2])), col="lightgrey", border=NA)
lines(dat[,1],m52b[,1],lwd=3,col="darkred"); lines(dat[,1],dat[,2],lwd=3,lty="dashed"); points(datr[1:15,1],datr[1:15,2],pch=19,cex=1.5)
text(10,30,"Matérn 5/2, N=15",pos=2);

plot(dat[,1],m32b[,1],type="l",lwd=3,ylim=c(-8,30),xlim=c(0.30,9.65),col="steelblue",xlab="x",ylab="f(x)"); 
polygon( c(dat[,1],rev(dat[,1])), c(m32b[,3], rev(m32b[,2])), col="lightgrey", border=NA)
lines(dat[,1],m32b[,1],lwd=3,col="steelblue"); lines(dat[,1],dat[,2],lwd=3,lty="dashed"); points(datr[1:15,1],datr[1:15,2],pch=19,cex=1.5)
text(10,30,"Matérn 3/2, N=15",pos=2);

########################################################################
# End of file
########################################################################

File: examplesForThesis-python/ch4-example-gp-regression.py
########################################################################
#
# Code to reproduce examples from the Licentiate's thesis:
# Sequential Monte Carlo for inference in nonlinear state space models
#
# written by
# Johan Dahlin ( johan.dahlin (at) liu.se
#
# Available from:
# http://users.isy.liu.se/en/rt/johda87/
#
# Copyright (c) 2014 Johan Dahlin
# 
# Run this code together with the corresponding R-file to reproduce:
#
# GP regression
# Example 4.6 in Section 4.4.1
#
########################################################################

import numpy as np
import pylab as pb
import GPy
import pandas

########################################################################
# Create the underlying function and evaluate it on a grid
########################################################################
x = np.linspace(0.,10.,500); x = x[:,None];
y = (3*cos(x/2) + sin(2+x/4) )**2;

# Select some random samples of the function and add some noise to the observations
xred = x[[ 20, 197, 317, 149,  89, 443,  53, 353, 395, 186, 292, 359, 326, 369,  99]]
yred = y[[ 20, 197, 317, 149,  89, 443,  53, 353, 395, 186, 292, 359, 326, 369,  99]] + np.sqrt(2) * np.random.normal(size=(15,1))
plot(x,y); plot(xred,yred,'ro')

########################################################################
# Fit the GP to the first 5 observations using the SE kernel
########################################################################
k1 = GPy.kern.bias(d) + GPy.kern.rbf(d, lengthscale=1);      
m1 = GPy.models.GPRegression(xred[0:5],yred[0:5],k1); m1.optimize();

# Evaluate the predictive distribution on a grid
Mup, ys2, up95, lo95 = m1.predict( np.array(x) ); 

# Export data to R for plotting
out = hstack((Mup,up95,lo95));
pandas.DataFrame(out).to_csv("../figures/ch4-example-gp-regression/ch4-example-gp-regression-rbf-10d.csv");

########################################################################
# Fit the GP to the first 5 observations using the Matern 3/2 kernel
########################################################################
k2 = GPy.kern.bias(d) + GPy.kern.Matern32(d, lengthscale=1);     
m2 = GPy.models.GPRegression(xred[0:5],yred[0:5],k2); m2.optimize();

# Evaluate the predictive distribution on a grid
Mup, ys2, up95, lo95 = m2.predict( np.array(x) ); 

# Export data to R for plotting
out = hstack((Mup,up95,lo95));
pandas.DataFrame(out).to_csv("../figures/ch4-example-gp-regression/ch4-example-gp-regression-m32-10d.csv");

########################################################################
# Fit the GP to the first 5 observations using the Matern 5/2 kernel
########################################################################
k3 = GPy.kern.bias(d) + GPy.kern.Matern52(d, lengthscale=1);     
m3 = GPy.models.GPRegression(xred[0:5],yred[0:5],k3); m3.optimize();

# Evaluate the predictive distribution on a grid
Mup, ys2, up95, lo95 = m3.predict( np.array(x) ); 

# Export data to R for plotting
out = hstack((Mup,up95,lo95));
pandas.DataFrame(out).to_csv("../figures/ch4-example-gp-regression/ch4-example-gp-regression-m52-10d.csv");

########################################################################
# Fit the GP to the 15 observations using the SE kernel
########################################################################
k1 = GPy.kern.bias(d) + GPy.kern.rbf(d, lengthscale=1);      
m1 = GPy.models.GPRegression(xred,yred,k1); m1.optimize();

# Evaluate the predictive distribution on a grid
Mup, ys2, up95, lo95 = m1.predict( np.array(x) ); 

# Export data to R for plotting
out = hstack((Mup,up95,lo95));
pandas.DataFrame(out).to_csv("../figures/ch4-example-gp-regression/ch4-example-gp-regression-rbf-20d.csv");

########################################################################
# Fit the GP to the 15 observations using the Matern 3/2 kernel
########################################################################
k2 = GPy.kern.bias(d) + GPy.kern.Matern32(d, lengthscale=1);     
m2 = GPy.models.GPRegression(xred,yred,k2); m2.optimize();

# Evaluate the predictive distribution on a grid
Mup, ys2, up95, lo95 = m2.predict( np.array(x) ); 

# Export data to R for plotting
out = hstack((Mup,up95,lo95));
pandas.DataFrame(out).to_csv("../figures/ch4-example-gp-regression/ch4-example-gp-regression-m32-20d.csv");

########################################################################
# Fit the GP to the 15 observations using the Matern 5/2 kernel
########################################################################
k3 = GPy.kern.bias(d) + GPy.kern.Matern52(d, lengthscale=1);     
m3 = GPy.models.GPRegression(xred,yred,k3); m3.optimize();

# Evaluate the predictive distribution on a grid
Mup, ys2, up95, lo95 = m3.predict( np.array(x) ); 

# Export data to R for plotting
out = hstack((Mup,up95,lo95));
pandas.DataFrame(out).to_csv("../figures/ch4-example-gp-regression/ch4-example-gp-regression-m52-20d.csv");

# Export data to R for plotting
out = hstack((x,y));
pandas.DataFrame(out).to_csv("../figures/ch4-example-gp-regression/ch4-example-gp-regression-data.csv");

# Export data to R for plotting
out = hstack((xred,yred));
pandas.DataFrame(out).to_csv("../figures/ch4-example-gp-regression/ch4-example-gp-regression-datared.csv");

########################################################################
# End of file
########################################################################
